design/20230726-cel-policy.md

```diff
@@ -0,0 +1,368 @@
+<!--
+This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
+-->
+
+# CEL expressions to approve or deny CertificateRequest
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories-optional)
+    - [X.509 DNS SAN in namespace domain](#x509-dns-san-in-namespace-domain)
+    - [SPIFFE ID identifying namespace](#spiffe-id-identifying-namespace)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Supported Versions](#supported-versions)
+- [Production Readiness](#production-readiness)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [CEL expressions with full decoded CSR variable](#cel-expressions-with-full-decoded-csr-variable)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged.
+
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented 
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+
+## Summary
+
+<!--
+This section is important for producing high-quality, user-focused
+documentation such as release notes.
+
+A good summary is probably around a paragraph in length.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+While approver-policy allows policies to include namespace selectors, this does not scale when attempting to
+enforce CSR attributes that must be a function of the CertificateRequest namespace (or other CertificateRequest fields).
+This proposal suggests an extension to the CertificateRequestPolicy API allowing policy authors to specify approval rules using
+[CEL expressions](https://github.com/google/cel-spec) containing variables with information from the CertificateRequest.
```

Comment:
Is there a reason to point to that instead of https://kubernetes.io/docs/reference/using-api/cel/?

Reply:
We could link to both. The Kubernetes docs give a good overview, but if you're interested in the details, I think you need to consult the detailed spec.

Reply:
Good point, thanks! I will change the link in my next round on this. Also thanks for the rewording suggestions! 👍